lib/evaluate/shape.h

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// GetShape() analyzes an expression and determines its shape, if possible,
+// representing the result as a vector of scalar integer expressions.
+
+#ifndef FORTRAN_EVALUATE_SHAPE_H_
+#define FORTRAN_EVALUATE_SHAPE_H_
+
+#include "expression.h"
+#include "tools.h"
+#include "type.h"
+#include "../common/indirection.h"
+#include <optional>
+#include <variant>
+
+namespace Fortran::evaluate {
+
+using ExtentType = SubscriptInteger;
+using ExtentExpr = Expr<ExtentType>;
+using MaybeExtent = std::optional<ExtentExpr>;
```

Comment:
There are lots of other places we use `Expr<SubscriptInteger>` and `std:optional<Expr<SubscriptInteger>>`. Should we use these same type aliases for all of them?

Reply:
Maybe.  I used these names here to try to clarify the declarations that use them.  But maybe they're a distraction.  The whole business with `SubscriptInteger` is a messy work-around to a problem that really should be fixed within the Fortran standard (default REAL is single-precision IEEE-754, default INTEGER has to be the same bit size as default REAL, and yet we want to be able to have arrays and loop bounds bigger than 32 bits).